### ISSUE TITLE
Leak Cycle in HTTPParser/HTTPServerRequest

### ISSUE THREAD

**Issue by carlbrown** [IBM-Swift/Kitura#902]
There's a retain cycle in base Kitura (see screenshot). 

Xcode Instruments trace files at https://github.com/carlbrown/KituraNoMiddlewareLeakSpike/tree/e521326dcbbfd1452dfde0e392e38f33ebf88782/LeakInfo).

**Comment by carlbrown** [comment]
![screen shot 2016-12-02 at 11 43 16 am](https://cloud.githubusercontent.com/assets/67206/20844571/1ba9e402-b886-11e6-9289-1f52b27827c4.png)

**Comment by youming-lin** [comment]
`HTTPParser` has a strong reference to an `HTTPParserDelegate` instance: https://github.com/IBM-Swift/Kitura-net/blob/master/Sources/KituraNet/HTTPParser/HTTPParser.swift#L34

`HTTPIncomingMessage` has a strong reference to an `HTTPParser` instance: https://github.com/IBM-Swift/Kitura-net/blob/f7218c3b373ec62eeb3008a4b20012848b24c143/Sources/KituraNet/HTTP/HTTPIncomingMessage.swift#L90

**Comment by carlbrown** [comment]
Yeah. The problem is that the way that HTTPParser manages its delegate, which is REALLY odd.
Look at https://github.com/IBM-Swift/Kitura-net/blob/master/Sources/KituraNet/HTTPParser/HTTPParser.swift#L154 

I've been unable to understand what's going on there.  I'm sure there's a reason for it, but it's not explained and certainly not intuitive.

Just making the reference `weak` just causes a `BAD_ACCESS` in that `getDelegate()` method.

**Comment by na-gupta** [comment]
@carlbrown If you're using Kitura >= 1.2.2, a big part of the leak may be because URLComponents that was added recently has a memory leak on linux(#901). I've just opened two PRs to replace that with URL which may help alleviate your problem.

But even if that did help, that retain cycle and the the HTTPParser delegate should probably be addressed separately.

**Comment by carlbrown** [comment]
@na-gupta This is all on Darwin (Instruments doesn't run on Linux), and it's tens of Kilobytes per page load, and I can't imagine that URLComponents could leak that much memory.

**Comment by shmuelk** [comment]
The retain loop between HTTPIncomingMessage and HTTPParser while unfortunate is on purpose.

HTTPParser is our wrapper for the C http_parser, which fires events as an HTTP request is parsed. In fact it doesn't even necessarily parse a request all in one shot, as the data may not have come in all in one shot.

HTTPParser in turn invokes functions on its delegate to notify it of the "status" of the parsing.

What is suppose to happen when we are done with a HTTPIncomingMessage (the super class of HTTPServerRequest) instance, is we call the HTTPIncomingMessage.release() function. That function clears the delegate in the HTTPParser. Also calls to HTTPIncomingMessage.freeHTTPParser() (called by HTTPIncomingMessage.release() ) clear the delegate in HTTPParser and then release the HTTPParser instance.

The real question here, is did we miss a release/free somewhere.

There have been attempts to make one of the references involved weak, that causes other failures.

**Comment by djones6** [comment]
As Shmuel says, this is a retain cycle that we have attempted to fix before.  Our current solution is to manually break the cycle when we know that we are finished with the parser.

The current solution works for workloads and (perhaps crucially) driving methods that I have studied to this point.  Having peppered the code with debug, it looks like the problem here is once again related to keepalive.  In the case where `IncomingSocketManager.removeIdleSockets()` finds a connection to clean up, I do not see the `HTTPIncomingMessage.release()` called (I haven't traced through why yet).  This situation never happens when driving load with `wrk`, because we always hit the maximum requests per connection, and perform an active cleanup at that point.

**Comment by shmuelk** [comment]
I think this may be partially related to various errors occurring, including the resetting of the socket by the HA. Those sockets may never get cleaned up right now.

In general if a socket doesn't process one hundred requests it will get cleaned up after being idle for one minute.

Part of the problem is related to our not marking certain errored sockets as idle.

Currently, we close sockets if a read error occurs. For write errors we don't do that.

**Comment by djones6** [comment]
Re. cleaning up expired keepalive connections:
```
IncomingSocketManager.removeIdleSockets()    (gets called upon a new connection)
-> IncomingSocketHandler.prepareToClose()
-> (probably) IncomingSocketHandler.close()
-> IncomingSockerHandler.handleCancel()    (via DispatchSourceRead.cancel() on OSX)
-> (probably) Socket.close()
```
Nowhere do we appear to call `IncomingHTTPSocketProcessor.close()`, or `ServerRequest.release()`, though we do set some flags on the processor during handleCancel().  Do we deliberately delay cleaning up the processor here?  It seems that after the socket is closed, or we detect it is already closed, we should be cleaning up the processor.

**Comment by shmuelk** [comment]
The cleanup happens once a minute or so, when new sockets come in.

**Comment by carlbrown** [comment]
These traces in instruments were captured on my laptop, not behind a HAProxy.

**Comment by carlbrown** [comment]
![count_v_mem](https://cloud.githubusercontent.com/assets/67206/20946935/780364ea-bbd2-11e6-8a7e-94e2cc056afb.png)

Here's the memory profile I'm seeing as a result of this and other leaks.

**Comment by shmuelk** [comment]
Have you tried that test using the Kitura-net branch issue_902?

That branch has a fix. It's awaiting testing to verify that it fixes the leak.

**Comment by carlbrown** [comment]
![memorytrace901902911](https://cloud.githubusercontent.com/assets/67206/20975825/4025e2de-bc66-11e6-9143-4df139fae297.png)
New trace with fixes for #901, #902 and #911 merged.  It's much better - the slope of the line is 1.4 instead of 4.1. (Units are the same as the previous picture in this thread)

**Comment by carlbrown** [comment]
OK, so this is interesting.  This is the same build (built from https://github.ibm.com/Carl-Brown1/SandboxCnC/blob/master/server/bin/build_for_901_902.sh) with the same test run against it (4 consecutive runs of https://github.com/carlbrown/KituraNoMiddlewareLeakSpike/blob/master/bin/parallel_user_test_run.sh run as `parallel_user_test_run.sh 10 10`) on the same server.

The difference is that for the blue dots, the test script was run on my laptop, and for the orange dots the test was run on the server itself. (I was having issues with the VPN dropping me, so I ran the test locally to the server, and noticed I got very different results).

There's no HA Proxy here, just laptop-> server but the network characteristics will be quite different. The localhost run is much faster, for one thing.

This implies to me that the leaks I'm seeing (at least now) vary by network conditions (maybe something isn't cleaning up after a dropped connection?).

![localhost_vs_laptop](https://cloud.githubusercontent.com/assets/67206/20983951/fe6d444a-bc83-11e6-97af-705f89666fab.png)
